Modify session handling

an-tao opened the pull request:
to resolve #567 #570

pmattione commented:
It's pretty late here so I'll test the code tomorrow.  Looks like it should work though, thanks!

pmattione commented:
Still getting set up to test.  You may want to consider changing the Session::get() function to returning T instead of const T& since holding a ref into there once the lock is dropped isn't safe.  Would break code but maybe you're helping them :)  Shouldn't break any functionality though unless someone is using mutable on the returned ref, but in that which case they should really be using this new modify function.  

Maybe also change the return type for operator[] from any& to any too.

an-tao commented:
> Still getting set up to test. You may want to consider changing the Session::get() function to returning T instead of const T& since holding a ref into there once the lock is dropped isn't safe. Would break code but maybe you're helping them :) Shouldn't break any functionality though unless someone is using mutable on the returned ref, but in that which case they should really be using this new modify function.
> 
> Maybe also change the return type for operator[] from any& to any too.

You are right, thanks.

an-tao commented:
It's pretty late here ^_-, I'll fix this tomorrow~

pmattione commented:
This function works great:
void modify(const std::string &key, const std::function<void(T &)> &handler)

Haven't tested the other modify(), but it looks fine to me.  Maybe you'd want to have that take a std::function too to match?  
Any maybe make the "using SessionMap" public to help with the function signature.  Not necessary though.  

Thanks!

an-tao commented:
> This function works great:
> void modify(const std::string &key, const std::function<void(T &)> &handler)
> 
> Haven't tested the other modify(), but it looks fine to me. Maybe you'd want to have that take a std::function too to match?
> Any maybe make the "using SessionMap" public to help with the function signature. Not necessary though.
> 
> Thanks!


Fixed and also Fixed #570 , please check it, thanks.

an-tao commented:
@pmattione Do you think we can merge this PR?

pmattione commented:
Appears to be working for me, and fixes the compiler warnings, thanks!
